src/lib/writer.ts

```diff
@@ -196,29 +196,63 @@ class Writer {
    *
    * @param buffer Contents of the octet string.
    */
-  writeVarOctetString (buffer: Buffer): void {
-    if (!Buffer.isBuffer(buffer)) {
+  writeVarOctetString (buffer: Buffer | Writer): void {
+    if (Buffer.isBuffer(buffer)) {
+      const MSB = 0x80
+
+      if (buffer.length <= 127) {
+        // For buffers shorter than 128 bytes, we simply prefix the length as a
+        // single byte.
+        this.writeUInt8(buffer.length)
+      } else {
+        // For buffers longer than 128 bytes, we first write a single byte
+        // containing the length of the length in bytes, with the most significant
+        // bit set.
+        const lengthOfLength = Math.ceil(buffer.length.toString(2).length / 8)
+        this.writeUInt8(MSB | lengthOfLength)
+
+        // Then we write the length of the buffer in that many bytes.
+        this.writeUInt(buffer.length, lengthOfLength)
+      }
+
+      this.write(buffer)
+    } else if (buffer instanceof Writer) {
+      buffer.prependLengthPrefix()
+      this.write(buffer)
+    } else {
       throw new TypeError('Expects a buffer')
     }
+  }
+
+  /**
+   * Write the length prefix for the current buffer at the beginning of the buffer.
+   */
+  prependLengthPrefix (): void {
+    let length = 0
+    for (let buffer of this.components) {
+      length += buffer.length
+    }
 
     const MSB = 0x80
 
-    if (buffer.length <= 127) {
+    if (length <= 127) {
       // For buffers shorter than 128 bytes, we simply prefix the length as a
       // single byte.
-      this.writeUInt8(buffer.length)
+      this.components.unshift(Buffer.from([length]))
     } else {
       // For buffers longer than 128 bytes, we first write a single byte
       // containing the length of the length in bytes, with the most significant
       // bit set.
-      const lengthOfLength = Math.ceil(buffer.length.toString(2).length / 8)
-      this.writeUInt8(MSB | lengthOfLength)
+      const lengthOfLength = Math.ceil(length.toString(2).length / 8)
+
+      const lengthBuffer = new Writer()
+      lengthBuffer.writeUInt8(MSB | lengthOfLength)
 
       // Then we write the length of the buffer in that many bytes.
-      this.writeUInt(buffer.length, lengthOfLength)
-    }
+      lengthBuffer.writeUInt(length, lengthOfLength)
 
-    this.write(buffer)
+      this.components.unshift(lengthBuffer.getBuffer())
```

Comment:
Instead of calling `getBuffer`, which will `Buffer.concat` (which copies), could this just unshift each component? (maybe a private `prepend` method)